HAL: Add PMC, EFC configuration APIs

tmplt opened the pull request:
Based upon @jamesmunns work, described in #14, but heavily refactored.

Closes #5.

tmplt committed:
manifest: increase internal feature flag granularity

Closes #11

jamesmunns committed:
add pmc, efc modules from @jamesmunns's work

Sourced from [0]; /vendor/asamx7x-hal.

[0] https://github.com/jamesmunns/same70-experiments/commit/58a3c34512f7b7877833edc3a1cc9b28f9d8941d

tmplt commented:
So far `defmt` usage has been removed/commented out. I'll go over the implementation in detail and likely apply some refactoring and generalization work so the implementation works for my use-case.

tmplt commented:
Depends on #15.

jamesmunns commented:
You might want to also take https://github.com/jamesmunns/same70-experiments/commit/58a3c34512f7b7877833edc3a1cc9b28f9d8941d, which fixed a couple things I found through testing.

tmplt commented:
Amended, thanks.

jamesmunns commented on hal/src/lib.rs:
```diff
@@ -145,3 +145,7 @@ pub use atsamv71q21b as target_device;
 pub mod serial;
 #[cfg(feature = "rev-b")]
 pub mod watchdog;
+#[cfg(feature = "rev-b")]
```
If you haven't seen the [`cfg-if`](https://docs.rs/cfg-if/latest/cfg_if/) crate, it's pretty handy for avoiding wordy cfgs like this.

michalfita commented:
Thanks @jamesmunns for the hint. From what I remember all MCUs in this series share peripherals implementations, only register allocation and number of instances varies.

Thanks @tmplt for bringing this in. I'll review over the weekend.

tmplt committed:
move feature checking to a build.rs instead

More features will be added in future commits that require explicit
configuration. Moved to build.rs to remove noise from HAL
implementation.

tmplt committed:
efc: find FWS via TryFrom impls

tmplt committed:
hal: apply fmt

tmplt commented:
I have briefly looked over the current PMC API and considered some changes, inspired by the in-progress `atsamd` clocking API. Instead of a monolithic `set_clocks` I think it would be beneficial if each clock is configured seperately. When configuring the clocks the hierarchy in Fig. 31-1 is likely to be consulted first; an equivalent API should make it easier to apply and debug any arising issues (i.e. being able to localize it to a single statement instead of requiring the end-user to dig into the HAL).

Here's is a draft of such an API application (not finalized):
```rust
    let pmc = pmc::new(device.PMC);

    // fundamental clock configuration (HCLK, FCLK, MCK)
    let mainck = pmc.get_mainck(MainClkConfig::Internal(12.to_Hz()))?;
    let plla = pmc.get_plla(PllaConfig { source: mainck, div: 1, mult: 4 })?;
    let (hclk, mck) = pmc.get_clocks(ClocksConfig { source: plla, pres: ClocksPrescaler::Pres4, div: MckDiv::Pres1 });

    // peripheral clock configuration
    let pck4 = pmc.get_peripheral(4, PeriphClkConfig { source: plla, pres: 1 })?;
```

martinmortsell commented on hal/Cargo.toml:
```diff
@@ -95,8 +95,8 @@ atsamv71q21b = { version = "0.21.0", path = "../pac/atsamv71q21b", optional = tr
 # Refer to §2 in the data sheet.
-v71 = []
-v70 = []
+v71 = ["vddio-3v"]
```
The Relevant Tables are 58-3 and 59-3.

tmplt replied:
Thanks, I'll add a reference.

tmplt replied:
Amended in d7e0f5756eae0fc13e69e47b29fce158f7a2d54f.

tmplt committed:
features: separate device-selected master feature from MCU info

device-selected is a meta-feature that is not described in §2.

tmplt commented on hal/build.rs:
```diff
@@ -0,0 +1,21 @@
+            "The HAL is built for a specific target device selected using a feature, but no such a feature was selected."
+        );
+    }
+
```
Checks for VDDIO are missing. A fail must also be added for ` (v70 || v71) && vddio-1v`.

tmplt replied:
Thinking over it now, I believe VDDIO information should be provided to the EFC module instead of being set statically. Consider the case when the MCU is dynamically downvolted. Using the API for this information is also much stronger than an environmental variable.

tmplt replied:
See https://github.com/atsams-rs/atsamx7x-rust/pull/16/commits/762e919bbf644113496becf1e15a2909a8410e4d.

martinmortsell committed:
Add Voltage Level Checks To build.rs

martinmortsell committed:
Fix Return Types

martinmortsell committed:
Add get_ Methods For All The Clocks (Except SLCK)

tmplt committed:
efc: set VDDIO level via ctor instead of feature

This change supports the theoretical case of non-constant VDDIO (for
E70/S70). An enum is also a much stronger association, rather than an
environmental variable (feature).

tmplt committed:
build: ensure a chip feature is enabled

That is, so the crate does not build if only device-selected is
enabled.

martinmortsell committed:
pmc: Add Methods For Configuring Plla And Mck

martinmortsell committed:
Merge branch 'feat/pmc' of https://github.com/GrepitAB/atsams70-rust into feat/pmc

martinmortsell committed:
pmc: Add Non-functional External Oscillator Support

martinmortsell committed:
pmc: Fix Main Clock Selection

martinmortsell committed:
pmc: Add PCK Configuration

martinmortsell committed:
Update Documentation

martinmortsell commented:
A basic implementation of the API @tmplt suggested above is now working.
I have successfully replaced the PAC level configuration I was using earlier.
However, the code does not provide any of the guard rails I would expect from a rust HAL.

michalfita commented on hal/src/pmc.rs:
```diff
@@ -0,0 +1,810 @@
+}
+
+impl PeripheralIdentifier {
+    pub fn supports_pmc_clocking(&self) -> Result<(), ()> {
```
Why `Result<>` and why not const?

martinmortsell committed:
pmc: Change Clock Source From Enum To Trait

martinmortsell committed:
Add get_slck() Method

tmplt committed:
pmc: apply formatting

tmplt committed:
boards/atsame70_xpro: init example skeleton

memory.x tail courtesy of [0].

[0] https://github.com/atsamd-rs/atsamd/commit/d7e054f49ea1efa006a07cc438d49ce6a9217312

tmplt committed:
boards/atsame70_xpro: doc how to erase fauly firmware

tmplt committed:
pmc: improve get_mainck

From my interpretation there there is only one single internal
oscillator (not counting the slow clocks) which can be overridden by a
single external oscillator, only that there are two switches that must
be flipped to use the external.

tmplt committed:
hal/pmc: refer to PMC as pmc, instead of periph

"pmc" is shorter than "periph", and make it clear what peripheral that
is being configured, if we in the future must configure more of them.

martinmortsell reviewed (commented):
No, I believe there are three settings.
The internal RC Oscillator, which is enabled by default.
The Main Crystal Oscillator, for when you're using an external crystal.
And the Main Crystal Oscillator in Bypass Mode, for when you're using an external oscillator.

The Xplained devkits have oscillators, that are powered externally and generate a clock signal without the Main Crystal Oscillator.
A crystal would not generate a clock signal by it self, but would need to be powered by the Main Crystal Oscillator.

See 30.5.4

- Main RC Oscillator
- Main Crystal Oscillator
- An External Clock signal (Bypass Mode)

martinmortsell commented:
> No, I believe there are three settings. The internal RC Oscillator, which is enabled by default. The Main Crystal Oscillator, for when you're using an external crystal. And the Main Crystal Oscillator in Bypass Mode, for when you're using an external oscillator.
> 
> The Xplained devkits have oscillators, that are powered externally and generate a clock signal without the Main Crystal Oscillator. A crystal would not generate a clock signal by it self, but would need to be powered by the Main Crystal Oscillator.

The extact same situation holds for the Slow Clock too.

tmplt commented:
I think I understand it now, with some help from §60.2.1:

1. Main RC Oscillator;
2. Main Crystal Oscillator in normal mode: crystal connected to XIN and XOUT and thus powered by the MCU; and
3. Main Crystal Oscillator in bypass mode: XIN connected to the clock signal (which can be a crystal not powered by the MCU, a function generator, or anything else that produces a valid clock signal) and XOUT disconnected.

tmplt committed:
hal/pmc: add main crystal osc in normal mode back

Changes made in 465f7d7 incorrectly assumed that there was only one
way to configure the main crystal oscillator.

tmplt committed:
hal/pmc: get_pllack: deconstruct PllaConfig

tmplt committed:
hal/pmc: refactor get_hclk

The datasheet documents two methods of selecting a clock for HCLK/MCK:
one for PLLs and one for the slow and main clock.

tmplt committed:
hal/pmc: record clock freq for MAINCK, PLLACK, and HCLK

We will be utilizing this information later down the line, for example
when configuring peripheral baud rates.

tmplt committed:
hal/pmc: get_hclk: set EFC wait states before switching clocks

tmplt committed:
hal/pmc: correctly configure UPLLCK and record its freq

tmplt committed:
hal/pmc: implement get_upllckdiv

UPLLCKDIV is derived from UPLLCK by either /1 or /2. Only the divider
is forwarded to HCLK and PCK, as per §31.3.

tmplt committed:
hal/pmc: use Hertz instead of Megahertz when logical

Some clocks will emit MHz (UPLLCK, MAINCK) while others will emit
Hz (PLLA - depending on dev/mult, and SLCK).

tmplt committed:
hal/pmc: record SLCK freq

tmplt committed:
hal/pmc: remove unnecessary Results

tmplt committed:
hal/pmc: handle unused_variables/dead_code warnings

tmplt committed:
atsame70_xpro: apply example clock hierarchy configuration

This will later be moved into a proper examples/*.rs when a GPIO API
is available.

tmplt committed:
hal/pmc: improve struct/enum docs

tmplt committed:
hal/pmc: remove unused enum

tmplt committed:
hal/pmc: get_slck: remove impl details from docs

tmplt committed:
hal/pmc: improve top-level docs

tmplt commented:
This is now ready for review. The documented example (which is a copy of the added `atsame70_xpro` example) has been tested on hardware. I also tested UPLL configuration, which unfortuately failed. I'll be looking into that in the meantime.

The module can probably be improved by some macro application, but I cannot prioritize that at the moment. More safety can also be added, such as ensuring that a clock that is in used it not turned off, but that too must wait for my part.

tmplt committed:
atsame70_xpro: expose UPLLCKDIV on PCK2, @ 2.4MHz

tmplt commented:
I must have spoken too soon: `atsame70_xpro` has been modified to enable the UPLL and expose it on PA03 via PCK2 @ 2.4MHz. Works as expected.

tmplt committed:
hal/pmc: globalize magic values

tmplt committed:
hal/pmc: apply formatting

tmplt committed:
hal/pmc: refactor SCLK configuration to match MAINCK

tmplt committed:
hal/pmc: remove outdated comment

michalfita commented on Cargo.lock:
```diff
@@ -728,6 +729,24 @@ dependencies = [
  "void",
 ]
 
+[[package]]
+name = "fugit"
+version = "0.3.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "f6d8595783d5ca52f0e9830036b3d24f359fae0fcc6bb5fde41f2dd82997cb58"
+dependencies = [
+ "gcd",
+]
+
+[[package]]
+name = "gcd"
+version = "2.1.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "f37978dab2ca789938a83b2f8bc1ef32db6633af9051a6cd409eff72cbaaa79a"
+dependencies = [
+ "paste",
```
I always thought the rule in Rust world is not to commit `Cargo.lock` for library crates.

tmplt replied:
Somehow slipped through; I'll remove it.

michalfita commented on hal/Cargo.toml:
```diff
@@ -93,136 +94,156 @@ atsamv71q21  = { version = "0.21.0", path = "../pac/atsamv71q21", optional = tru
 
 [features]
+# Internal-only feature flags; do not set directly.
+# Refer to §2 in the data sheet.
+## Refer to §2.
```
? Why refer doubled here?

tmplt replied:
Good catch.

michalfita commented on boards/atsame70_xpro/src/main.rs:
```diff
@@ -0,0 +1,94 @@
+            Efc::new(ctx.device.EFC, VddioLevel::V3)
+        };
+
+        // Configure the clock hierarchy
+        {
+            use hal::pmc::{
+                HostClockConfig, MainCkSource, MckDivider, MckPrescaler, Megahertz, PckId,
+                PllaConfig, Pmc, UpllDivider,
+            };
+
+            let mut pmc = Pmc::new(ctx.device.PMC);
+            let mainck = pmc
+                .get_mainck(MainCkSource::ExternalBypass(Megahertz::from_raw(12)))
+                .unwrap();
+            let _plla = pmc
+                .get_pllack(PllaConfig { div: 1, mult: 8 }, &mainck)
+                .unwrap();
+            let _hclk = pmc
+                .get_hclk(
+                    HostClockConfig {
+                        pres: MckPrescaler::CLK_1,
+                        div: MckDivider::EQ_PCK,
+                    },
+                    &mainck,
+                    &mut efc,
+                )
+                .unwrap();
+
+            let upllck = pmc.get_upllck(&mainck, &mut ctx.device.UTMI).unwrap();
+            let upllckdiv = pmc.get_upllckdiv(&upllck, UpllDivider::Div2);
+            let _pck2 = pmc.get_pck(&upllckdiv, 100 - 1, PckId::Pck2); // @ 2.4MHz
+        }
```
Wouldn't be more readable if these are separate functions?

tmplt replied:
I'm not sure what you mean. What functions here could be further separated?

michalfita replied:
The whole function is long but seems to have clear separation into stages. To improve readability they could be shorter private functions with descriptive names.

michalfita commented on hal/src/efc.rs:
```diff
@@ -0,0 +1,108 @@
+///
+/// Refer to §58 and §59.
+#[derive(PartialEq)]
+pub enum VddioLevel {
+    /// VDDIO = 3.3V, typical
+    V3,
+    /// VDDIO = 1.7V, minimal
+    V1,
+}
```
Wouldn't `V3v3` and `V1v7` more _accurate_ in places where code is read?

tmplt replied:
Good idea; those are much better names.

michalfita commented on hal/src/efc.rs:
```diff
@@ -0,0 +1,108 @@
+}
+
+impl FlashWaitStates {
+    pub fn calculate(freq: Megahertz, vddio: &VddioLevel) -> Result<Self, PmcError> {
+        #[cfg(any(feature = "v70", feature = "v71"))]
+        if vddio == &VddioLevel::V1 {
+            // V70/V71 must be driven with VDDIO = 3.3V, typical
+            return Err(PmcError::InvalidConfiguration);
+        }
+
+        Self::fws_from_freq(freq, vddio)
+    }
```
Shouldn't we prevent the invalid configuration at compile time?

tmplt replied:
We should. I had not played around with the type system before I began work on #18. I'll review this module and see what can be made static.

michalfita commented on hal/src/pmc.rs:
```diff
@@ -0,0 +1,808 @@
+                    w
+                });
+
+                // TODO hande note for MOSCRCF unhandled (p. 276;
+                // first table, second row)
```
Can I suggest one of _commercial_ practices that I like? Turn each TODO into an issue, then point the issue ID with the comment like: `// TODO #333: Handle note for MOSCRF` - I've learnt TODO comments quickly gets neglected and later in time confuse other contributors. With issue in place there's more urge to actually do the stuff.

tmplt replied:
Yep, let's do that.

michalfita commented on hal/src/pmc.rs:
```diff
@@ -0,0 +1,808 @@
+                });
+                while self.pmc.pmc_sr.read().moscsels().bit_is_clear() {}
+
+                // TODO check MAINCK frequency (§31.17; step 5).
```
Ditto.

michalfita commented on hal/src/pmc.rs:
```diff
@@ -0,0 +1,808 @@
+    }
+
+    /// Configures PLLACK and returns a corresponding clock token.
+    /// This method corresponds to Step 6 of 31.17 Recommended Programming Sequence.
+    pub fn get_pllack<SRC: PllaSource>(
```
Would be nice to equip docs of these API functions with some examples?

tmplt replied:
Do you mean an example for each `get_`-function?

michalfita replied:
Yes.

michalfita commented on hal/src/pmc.rs:
```diff
@@ -0,0 +1,808 @@
+        if div == 0 || div > 127 {
+            return Err(PmcError::InvalidConfiguration);
+        }
+        // TODO: Ensure valid requested output frequency.
```
Ditto.

michalfita reviewed (commented):
Looks good in overall, but I have a few suggestions if you don't mind.

tmplt commented:
Thanks for the review, @michalfita. I am busy with internal development of this HAL at the moment, but aim to continue the upstreaming process here when time is available. It can take a while until then.

michalfita commented:
> It can take a while until then.

You @martinmortsell and @jamesmunns are probably only users of this work ATM, so it's fine. However, it would be nice to release it at some point.

tmplt commented:
Branch is now deprecated. Closing. Refer to https://github.com/atsams-rs/atsamx7x-rust/discussions/24.
